File: 2017/kotlin/day11.kt
import java.io.File
import kotlin.math.abs; import kotlin.math.max; import kotlin.math.sign

val move = mapOf("n" to Pair(0, 1), "ne" to Pair(1, 0), "se" to Pair(1, -1), "s" to Pair(0, -1), "sw" to Pair(-1, 0), "nw" to Pair(-1, 1))

fun hexManhattanDist(pos: Pair<Int, Int>): Int = if (pos.first.sign == pos.second.sign) abs(pos.first) + abs(pos.second) else max(abs(pos.first), abs(pos.second))

fun main() {
    val steps = File("../__in/11.txt").readText().trim().split(",")
    var pos = Pair(0, 0)
    var maxDist = 0
    
    for (step in steps) {
        pos += move[step]!!
        with(hexManhattanDist(pos)) { if (maxDist < this) maxDist = this }
    }
    
    println("Part 1: ${hexManhattanDist(pos)}")
    println("Part 2: $maxDist")
}

File: 2017/lua/day11.lua
stuff = require("stuff")

--[[    a       a.         _        let's invent hexagonal wheel for fun!
        |       |  `b =   .c    <-- axis are set such as (1, 0, 0) + (0, 1, 0) = (0, 0, -1), and so on
       .o.      o      o'           also any triplet is reducible to (+n1, -n2, 0) or set with more zeroes,
    c'     `b                       and reduced form gives the Manhattan distance just like Carthesian (x, y). ]]

move = {n = {1, 0, 0}, ne = {0, 0, -1}, se = {0, 1, 0}, s = {-1, 0, 0}, sw = {0, 0, 1}, nw = {0, -1, 0}}

function step(pos, dir)
    for i = 1, 3 do pos[i] = pos[i] + move[dir][i] end
    return pos end

function reduce(t)
    local tmp = {}; for i = 1, 3 do table.insert(tmp, t[i]) end
    table.sort(tmp); local delta = tmp[2]
    
    for i = 1, 3 do tmp[i] = t[i] - delta end
    return tmp end

function hexmd(t) return math.abs(t[1] - t[2] - t[3]) end

input = stuff.str2strtable(io.open("11.txt", "r"):read())

pos, max = {0, 0, 0}, 0
for _, v in ipairs(input) do
    pos = reduce(step(pos, v))
    local d = hexmd(pos); if max < d then max = d end
end

print(string.format("Part 1: %d", hexmd(pos)))
print(string.format("Part 2: %d", max))